Configuration adjustment in Datafusion

metesynnada opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #.

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

metesynnada committed:
Initial but not completely work like proto

metesynnada committed:
Before proto is handled

metesynnada committed:
Merge remote-tracking branch 'upstream/main' into configuration

metesynnada committed:
Update listing_table_factory.rs

metesynnada committed:
Before proto 2

metesynnada committed:
Minor adjustments

Weijun-H committed:
feat: support `FixedSizeList` Type Coercion (#9108)

* support FixedSizeList Type Coercion

* add allow null type coercion parameter

* support null column in FixedSizeList

* Add test

* Add tests for cardinality with fixed size lists

* chore

* fix ci

* add comment

* Fix array_element function signature

* Remove unused imports and simplify code

* Fix array function signatures and behavior

* fix conflict

* fix conflict

* add tests for FixedSizeList

* remove unreacheable null check

* simplify the code

* remove null checking

* reformat output

* simplify code

* add tests for array_dims

* Refactor type coercion functions in datafusion/expr module

jonahgao committed:
feat: expand `unnest`  to accept arbitrary single array expression (#9342)

* feat: expand `unnest`  to accept any single array expression

* unnest null

* review feedback

metesynnada committed:
Merge branch 'main' into configuration

codonnell committed:
fix: flight examples (#9335)

* fix: downgrade tonic for arrow compatibility

Tonic 0.10 and 0.11 are not API compatible.
Arrow 50 depends on tonic 0.10, and datafusion must match that dependency for compatibility reasons.

* feat: make nested examples runnable

cargo run --example doesn't support nested examples. Nested examples need an explicit block to be runnable.

* fix: fix custom catalog typo and formatting

* docs: add note about upgrading tonic with arrow

* ci: add cargo check for all examples

tshauck committed:
docs: update parquet_sql_multiple_files.rs with a relative path ex (#9310)

* docs: update parquet_sql_multiple_files.rs with a relative path ex

* style: run cargo fmt

* docs: update comment

* docs: better

tshauck committed:
tests: add tests for writing hive-partitioned parquet (#9316)

* tests: adds tests associated with #9237

* style: clippy

guojidan committed:
feature: support nvl(ifnull) function (#9284)

* feature: support nvl(ifnull) function

* add sqllogictest

* add docs entry

* Update docs/source/user-guide/sql/scalar_functions.md

Co-authored-by: Jonah Gao <jonahgao@msn.com>

* fix some code

* fix docs

---------

Co-authored-by: Jonah Gao <jonahgao@msn.com>

yyy1000 committed:
Move abs to datafusion_functions (#9313)

* feat: move abs to datafusion_functions

* fix proto

* fix proto

* fix CI vendored code

* Fix proto

* add support type

* fix signature

* fix typo

* fix test cases

* disable a test case

* remove old code from math_expressions

* feat: add test

* fix clippy

* use unknown for proto

* fix unknown proto

mustafasrepo commented on datafusion/common/src/config.rs:
```diff
@@ -926,9 +928,30 @@ config_field!(String);
 config_field!(bool);
 config_field!(usize);
 config_field!(f64);
-config_field!(u8);
```
Why we need to update default implementation for u8?

mustafasrepo commented on datafusion/common/src/config.rs:
```diff
@@ -1061,3 +1084,484 @@ macro_rules! extensions_options {
+        let (key, rem) = key.split_once('.').unwrap_or((key, ""));
+        assert_eq!(key, "kafka");
+        self.properties.insert(rem.to_owned(), value.to_owned());
+        println!("key: {}, value: {}", rem, value);
```
This print is leftover

mustafasrepo commented on datafusion/common/src/config.rs:
```diff
@@ -1061,3 +1084,484 @@ macro_rules! extensions_options {
+            .set("format.parquet.write_batch_size", "10")
+            .unwrap();
+        assert_eq!(table_config.format.parquet.global.write_batch_size, 10);
+        table_config.set("kafka.bootstrap.servers", "mete").unwrap();
```
This variable maybe leftover

mustafasrepo commented on datafusion/core/src/dataframe/mod.rs:
```diff
@@ -1171,28 +1156,28 @@ impl DataFrame {
             FileType::CSV,
-            options.partition_by,
             copy_options,
+            Default::default(),
```
I think, you can use `HashMap::new()` instead to be more explicit

mustafasrepo commented on datafusion/core/src/dataframe/mod.rs:
```diff
@@ -1224,21 +1210,28 @@ impl DataFrame {
             FileType::JSON,
-            options.partition_by,
             copy_options,
+            Default::default(),
```
Similar to comment above

mustafasrepo commented on datafusion/core/src/dataframe/parquet.rs:
```diff
@@ -53,30 +50,30 @@ impl DataFrame {
             FileType::PARQUET,
-            options.partition_by,
             copy_options,
+            Default::default(),
```
Similar to comment above

crepererum committed:
refactor: `SchemaProvider::table` can fail (#9307)

mustafasrepo commented on datafusion/core/src/datasource/file_format/parquet.rs:
```diff
@@ -540,6 +548,8 @@ async fn fetch_statistics(
 pub struct ParquetSink {
     /// Config options for writing data
     config: FileSinkConfig,
+    ///
```
Dcostring is missing

mustafasrepo committed:
Update headers

mustafasrepo commented on datafusion/core/src/physical_planner.rs:
```diff
@@ -602,14 +592,13 @@ impl DefaultPhysicalPlanner {
 
                     let sink_format: Arc<dyn FileFormat> = match file_format {
-                        FileType::CSV => Arc::new(CsvFormat::default()),
+                        FileType::CSV => Arc::new(CsvFormat::default().with_options(format_options.csv.clone())),
```
I think you can use `CsvFormat::new()` instead of `CsvFormat::default()`

mustafasrepo commented on datafusion/expr/src/logical_plan/dml.rs:
```diff
@@ -39,46 +39,29 @@ pub struct CopyTo {
-    /// Holds FileTypeWriterOptions directly provided
-    WriterOptions(Box<FileTypeWriterOptions>),
+    pub format_options: FormatOptions,
+    ///
```
Docstring is missing

metesynnada committed:
Update copy.slt

mustafasrepo committed:
Merge branch 'configuration' of https://github.com/synnada-ai/datafusion-upstream into configuration

mustafasrepo committed:
Add new test,

guojidan committed:
fix write_partitioned_parquet_results bug (#9360)

DDtKey committed:
fix: use `JoinSet` to make spawned tasks cancel-safe (#9318)

* fix: use `JoinSet` to make spawned tasks cancel-safe

* feat: drop `AbortOnDropSingle` and `AbortOnDropMany`

* style: doc lint

* fix: ordering of the tasks in `RepartitionExec`

* fix: replace spawn_blocking with JoinSet

* style: disallow spawn methods

* fixes: preserve ordering of tasks

* style: allow spawning in tests

* chore: exclude clippy.toml from rat

* chore: typo

* feat: introduce `SpawnedTask`

* revert outdated comment

* switch to SpawnedTask missed outdated part

* doc: improve reason for disallowed-method

dependabot[bot] committed:
Update nix requirement from 0.27.1 to 0.28.0 (#9344)

Updates the requirements on [nix](https://github.com/nix-rust/nix) to permit the latest version.
- [Changelog](https://github.com/nix-rust/nix/blob/master/CHANGELOG.md)
- [Commits](https://github.com/nix-rust/nix/compare/v0.27.1...v0.28.0)

---
updated-dependencies:
- dependency-name: nix
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Omega359 committed:
Replace usages of internal_err with exec_err where appropriate (#9241)

* internal_err! -> exec_err!

* fmt updates.

* Updated error macro from exec_err! to not_impl_err! for all unsupported type errors.

* Revert "Updated error macro from exec_err! to not_impl_err! for all unsupported type errors."

This reverts commit fe0517a7ca0c0c6e7fa2a61b0bf321d45185854c.

* Updated a few instances of internal_err missed in previous audit.

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

metesynnada committed:
Passes SLT tests

mobley-trent committed:
feat : Support for deregistering user defined functions (#9239)

* Initial commit

* Updated mod.rs - Docstrings, Initial test

* Updated mod.rs - Fixed udf test

* Added udaf test, Updated udf test

* Added test for udwf

* Linting with rustfmt

* Update datafusion/core/src/execution/context/mod.rs

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* Moved tests to core/tests/user_defined

* fix fmt

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

guojidan committed:
fix return type (#9357)

SteveLauC committed:
refactor: move acos() to function crate (#9297)

SteveLauC committed:
docs: put flatten in top fn list (#9376)

edilc committed:
Update list_to_string alias to point to array_to_string (#9374)

metesynnada committed:
Update csv.rs

metesynnada committed:
Before trying char

Lordworms committed:
feat: issue_9285: port builtin reg function into datafusion-function-* (1/3 regexpmatch) (#9329)

* feat: issue_9285: port builtin reg function into datafusion-function-* crate (1/3: RegexpMatch part)

* fix fmt

* refact

* modify test

* fix msrv verify problem

* port test and delete useless lines

metesynnada committed:
Fix u8 handling

jonahgao committed:
Add test to verify issue #9161 (#9265)

* Add test to verify issue #9161

* fmt

crepererum committed:
refactor: fix error macros hygiene (#9366)

jonahgao committed:
feat: support for defining ARRAY columns in `CREATE TABLE` (#9381)

SteveLauC committed:
fix: panic in isnan() when no args are given (#9377)

* fix: panic in isnan() when no args are given

* test: add sqllogictest for abs/acos/isnan

jonahgao committed:
feat: support `unnest` in FROM clause (#9355)

* feat: support `unnest` in FROM clause

* add test

* remove incorrect test

guojidan committed:
feat: support nvl2 function (#9364)

* feat: support nvl2 function

* fix signature && test case

SteveLauC committed:
refactor: move asin() to function crate (#9379)

korowa committed:
fix: using test data sample for catalog example (#9372)

* fix: test data sample for catalog example

* fix ci script comment

* review comments

Tangruilin committed:
delete tail space (#9386)

Signed-off-by: tangruilin <tang.ruilin@foxmail.com>

alamb committed:
Run cargo-fmt on `datafusion-functions/core` (#9367)

* Run cargo-fmt on datafusion-functions/core

* fmt nvl2

mustafasrepo committed:
Cache common plan properties to eliminate recursive calls in physical plan (#9346)

* Initial commit

* Update comments

* Review Part 1

* Minor changes

* Delete docs.yaml

* use create_cache_convention

* Address reviews

* Address reviews

* Update datafusion/physical-plan/src/lib.rs

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* Update comments

* Move properties to another trait.

* Final review

* Resolve linter errors

* Bring docs yaml

* Final reviews and cleanups

---------

Co-authored-by: Mehmet Ozan Kabak <ozankabak@gmail.com>
Co-authored-by: metesynnada <100111937+metesynnada@users.noreply.github.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

alamb committed:
Run cargo-fmt on all of `datafusion-functions` (#9390)

* Use modules directly in functons

* cargo fmt

Lordworms committed:
feat: issue #9224 substitute tlide in table path (#9259)

* substitute tilde in file path

* change test

* disable test on windows

* fix typo

Lordworms committed:
port range function and change gen_series logic (#9352)

* port range function and change gen_series logic

* fix failed test

* delete useless and add tests

* change parameter

* fix document

* remove useless

* change doc

* delete space

mustafasrepo committed:
[MINOR]: Generate physical plan, instead of logical plan in the bench test (#9383)

* initial commit

* Add logical plan all benchmark

metesynnada committed:
Merge remote-tracking branch 'upstream/main' into configuration

metesynnada committed:
Update according to review

Tangruilin committed:
[task #8987]add_to_date_function (#9019)

Signed-off-by: tangruilin <tang.ruilin@foxmail.com>

alamb committed:
Minor: clarify performance in docs for `ScalarUDF`, `ScalarUDAF` and `ScalarUDWF` (#9384)

waynexia committed:
feat: replace std Instant with wasm-compatible wrapper (#9189)

* feat: replace std Instant with wasm-compatible wrapper

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

* update cli's lock file

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

* wrap instant in datafusion_common

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

* write unit test

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

* update lock file

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

* format toml

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

* Update datafusion/wasmtest/src/lib.rs

* add clippy lint to prevent from using std Instant

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

* rename mod name from wasm to instant

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

* fix existing std instant reference

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

* Update datafusion/common/src/instant.rs

Co-authored-by: Artem Medvedev <i@ddtkey.com>

---------

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>
Co-authored-by: Artem Medvedev <i@ddtkey.com>

Jefffrey committed:
Uplift keys/dependencies to use more workspace inheritance (#9293)

* Uplift keys/dependencies to use more workspace inheritance

* Update comment

* Whitespace

alamb committed:
Improve documentation for ExecutionPlanProperties, use consistent field name (#9389)

* Improve documentation for ExecutionPlanProperties, use consistent field name

* Add impl ExecutionPlanProperties for &dyn ExecutionPlan

devinjdangelo committed:
Doc: Workaround for Running cargo test locally without signficant memory (#9402)

* document cargo test workaround

* Update docs/source/contributor-guide/index.md

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

jonahgao committed:
feat: support `unnest` with additional columns (#9400)

* feat: support `unnest` with additional columns

* add test from issue

* add test to verify preserve_nulls

* update test

* fix name conflicts

jonahgao committed:
Minor: improve the display name of `unnest` expressions (#9412)

2010YOUY01 committed:
Minor: Move function signature check to planning stage (#9401)

* check signature for udf-based function impl

* clippy

dependabot[bot] committed:
chore(deps): update substrait requirement from 0.24.0 to 0.25.1 (#9406)

Updates the requirements on [substrait](https://github.com/substrait-io/substrait-rs) to permit the latest version.
- [Release notes](https://github.com/substrait-io/substrait-rs/releases)
- [Changelog](https://github.com/substrait-io/substrait-rs/blob/main/CHANGELOG.md)
- [Commits](https://github.com/substrait-io/substrait-rs/compare/v0.24.0...v0.25.1)

---
updated-dependencies:
- dependency-name: substrait
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

SteveLauC committed:
docs: update contributor guide (migration to sqllogictest is done) (#9408)

Omega359 committed:
Move the to_timestamp* functions to datafusion-functions (#9388)

* WIP

* WIP

* cargo fmt updates.

* Migrate to_timestamp* functions to new functions crate.

* update to allow wasm run to complete.

* Fix expr_api example

* cargo fmt.

* Update datafusion/core/tests/simplification.rs

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* Revert changes to sql in tests to restore to_timestamp(..) calls.

* Rust fmt apparently can't make up it's mind.

* Updates for merge and moving some expression simplifier tests from optimizer to core/tests.

* Revert test changes

* Update datafusion/core/tests/simplification.rs

* Move cast_column to `ColumnarValue::cast_to`

* Remove datafusion-physical-expr dependency

* Move dependencies to dev

* Fix merge conflict by migrating to_date to new functions module alongside to_timestamp.

* Cargo lock update from merge.

* Add missing licenses.

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

jayzhan211 committed:
Minor: Support LargeList List Range indexing and fix large list handling in ConstEvaluator  (#9393)

* fix largelist

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* support large list for list range

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

---------

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

pinarbayata committed:
NEW Logo (#9385)

* logo images are added

* update guide

* add new image

* Update readme

* Update images

* Move old logo related images to under old_logo folder

---------

Co-authored-by: Mustafa Akur <mustafa.akur@synnada.ai>

yyy1000 committed:
Handle serde for ScalarUDF (#9395)

caicancai committed:
Minior: Add negative tests about sqrt function (#9426)

mustafasrepo committed:
Move SpawnedTask from datafusion_physical_plan to new `datafusion_common_runtime` crate (#9414)

* Initial commit

* Tmp

* Remove clippy ignores

* Move SpawnedTask to under new crate

* Minor changes

* re-export from core

andygrove committed:
Re-export datafusion-functions-array (#9433)

PsiACE committed:
Support LargeList for ListIndex (#9424)

Signed-off-by: Chojan Shang <psiace@apache.org>

Weijun-H committed:
move ArrayDims, ArrayNdims and Cardinality to datafusion-function-crate (#9425)

* Update array functions and remove ArrayDims and Cardinality

* move ArrayNdims function

* add roundtrip tests

SteveLauC committed:
refactor: make instr() an alias of strpos() (#9396)

MohamedAbdeen21 committed:
Add test case for invalid tz in timestamp literal (#9429)

* Add test case for invalid tz in timestamp literal

* add a separate unit test

---------

Co-authored-by: Mohamed Abdeen <mohamed.abdeen@paytabs.com>

alamb committed:
Minor: simplify call (#9434)

comphead committed:
Support IGNORE NULLS for LEAD window function (#9419)

* IGNORE NULLS support for LEAD

* fix

* fix

* fix

jackwener committed:
fix sqllogicaltest result (#9444)

my-vegetable-has-exploded committed:
docs: rm duplicate words. (#9449)

jackwener committed:
minor: fix cargo clippy some warning (#9442)

Lordworms committed:
port regexp_like function and port related tests (#9397)

* port regexp_like function and port related tests

delete useless test

fix chores

change regexp_like

change lock

change format

fix typo

change prost

remove unused

adding usage

adding dependency

adding dependency

dep

* adding tests

* remove useless

* remove unused

* remove unused

* change dependency structure

* format toml

JasonLi-cn committed:
fix: sort_batch function unsupported mixed types with list (#9410)

* fix: sort_batch function unsupported mixed types with list

* chore: add some SQL level end to end tests

* refactor: Use RowConverter only when sorting multiple columns that contain a List type

* Update datafusion/physical-plan/src/sorts/sort.rs

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

DDtKey committed:
refactor: add `join_unwind` to `SpawnedTask` (#9422)

* refactor: add `join_unwind` to `SpawnedTask`

In order to remove duplication of these handlers it seems logical to have such method.

I thought to add this logic to `join` but there are methods with additional logic

* docs: improve join_unwind comments

* docs: improve join_unwind comments

mustafasrepo committed:
Ignore null LEAD support for small batch sizes. (#9445)

* IGNORE NULLS support for LEAD

* fix

* fix

* fix

* Add lead support when pruned

* Update lead_lag.rs

Minor changes

* Fix formatting

---------

Co-authored-by: comphead <comphead@ukr.net>
Co-authored-by: Mustafa Akur <akurmustafa@gmail.com>

jonahgao committed:
fix: casting to ARRAY types failed (#9441)

* fix: casting to ARRAY types failed

* use DataType::new_list

korowa committed:
fix: reading from partitioned `json` & `arrow` tables (#9431)

* fix partitioned table reading for json

* wildcard projection test for csv partitioned table

* review comments

* fix partitioned arrow tables reading

JasonLi-cn committed:
feat: Support `EscapedStringLiteral`, update sqlparser to `0.44.0` (#9268)

* feat: upgrade sqlparser from 0.43.0 to 0.44.0

* chore: update Cargo.lock

comphead committed:
Minor: fix LEAD test description (#9451)

* fix LEAD test description

* fix test description

peter-toth committed:
Consolidate `TreeNode` transform and rewrite APIs (#8891)

* refactor `TreeNode::rewrite()`

* use handle_tree_recursion in `Expr`

* use macro for transform recursions

* fix api

* minor fixes

* fix

* don't trust `t.transformed` coming from transformation closures, keep the old way of detecting if changes were made

* rephrase todo comment, always propagate up `t.transformed` from the transformation closure, fix projection pushdown closure

* Fix `TreeNodeRecursion` docs

* extend Skip (Prune) functionality to Jump as it is defined in https://synnada.notion.site/synnada/TreeNode-Design-Proposal-bceac27d18504a2085145550e267c4c1

* fix Jump and add tests

* jump test fixes

* fix clippy

* unify "transform" traversals using macros, fix "visit" traversal jumps, add visit jump tests, ensure consistent naming `f` instead of `op`, `f_down` instead of `pre_visit` and `f_up` instead of `post_visit`

* fix macro rewrite

* minor fixes

* minor fix

* refactor tests

* add transform tests

* add apply, transform_down and transform_up tests

* refactor tests

* test jump on both a and e nodes in both top-down and bottom-up traversals

* better transform/rewrite tests

* minor fix

* simplify tests

* add stop tests, reorganize tests

* fix previous merges and remove leftover file

* Review TreeNode Refactor (#1)

* Minor changes

* Jump doesn't ignore f_up

* update test

* Update rewriter

* LogicalPlan visit update and propagate from children flags

* Update tree_node.rs

* Update map_children's

---------

Co-authored-by: Mustafa Akur <mustafa.akur@synnada.ai>

* fix

* minor fixes

* fix f_up call when f_down returns jump

* simplify code

* minor fix

* revert unnecessary changes

* fix `DynTreeNode` and `ConcreteTreeNode` `transformed` and `tnr` propagation

* introduce TransformedResult helper

* fix docs

* restore transform as alias to trassform_up

* restore transform as alias to trassform_up 2

* Simplifications and comment improvements (#2)

---------

Co-authored-by: Berkay Şahin <124376117+berkaysynnada@users.noreply.github.com>
Co-authored-by: Mustafa Akur <mustafa.akur@synnada.ai>
Co-authored-by: Mehmet Ozan Kabak <ozankabak@gmail.com>

Lordworms committed:
Support `Date32` arguments for `generate_series` (#9420)

* Supporting Date type for range and generate_series

* fix lock

* change lock

* optimize code and tests

* recover lock

* change test

* solve conflicts

* resolve conflicts

Lordworms committed:
Minor: change doc for range (#9455)

* change doc for range

* fix format

* adding examples

* fmt

* more explicit

metesynnada committed:
Passing tests

metesynnada committed:
passing tests with proto

metesynnada committed:
Cargo fix

Weijun-H committed:
Add missing doc index (#9462)

comphead committed:
update bigdecimal version (#9471)

dependabot[bot] committed:
chore(deps): update base64 requirement from 0.21 to 0.22 (#9446)

* chore(deps): update base64 requirement from 0.21 to 0.22

Updates the requirements on [base64](https://github.com/marshallpierce/rust-base64) to permit the latest version.
- [Changelog](https://github.com/marshallpierce/rust-base64/blob/master/RELEASE-NOTES.md)
- [Commits](https://github.com/marshallpierce/rust-base64/compare/v0.21.0...v0.22.0)

---
updated-dependencies:
- dependency-name: base64
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

* update datafusion-cli.lock

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

Lordworms committed:
Port regexp_replace functions and related tests (#9454)

* Port regexp_replace functions and related tests

* porting tests

* delete files

* change config

* adding dependency

* change tests

* fix

* optimize code

* remove unused

Omega359 committed:
Update contributor guide with updated scalar function howto (#9438)

* Update add new scalar function documentation.

* Improved wording for features, added missing crypto and unicode links.

universalmind303 committed:
feat: add support for fixed list wildcard in type signature (#9312)

* feat: add support for fixed list wildcard in type signature

* fmt

* clippy

* update tests

* switch coercing

* update logic

* fix test

* add tests to make sure it cant coerce diff sizes

* add test for same type

jayzhan211 committed:
Add a `ScalarUDFImpl::simplfy()` API, move `SimplifyInfo` et al to datafusion_expr (#9304)

* first draft

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* clippy

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* add comments

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* move to optimize rule

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* cleanup

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* fix explain test

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* move to simplifier

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* pass with schema

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* fix explain

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* fix doc

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* move to expr

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* change simplify signature

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* cleanup

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* cleanup

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* fix doc

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* fix doc

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* Update datafusion/expr/src/udf.rs

* Add backwards compatibile uses, inline FunctionSimplifier, rename to ExprSimplifyResult

* Remove DFSchema from SimplifyInfo

* Avoid requiring argument copies

* Improve docs

* fix link

* fix doc test

* Update datafusion/physical-expr/src/lib.rs

* Change example simplify to always simplify its argument

* Clarify comment

---------

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

huaxingao committed:
Implement IGNORE NULLS for FIRST_VALUE (#9411)

* Implement IGNORE NULLS for FIRST_VALUE

* fix style

* fix clippy error

* fix clippy error

* address comments

* fix error

* add test to aggregate.slt

* address comments

* Trigger Build

* Add one additional column in order by to ensure a deterministic order in the output

---------

Co-authored-by: Huaxin Gao <huaxin.gao@apple.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

milenkovicm committed:
Add plugable handler for `CREATE FUNCTION` (#9333)

* Add plugable function factory

* cover `DROP FUNCTION` as well ...

... partially, as `SessionState` does not expose
unregister_udf at the moment.

* update documentation

* fix doc test

* Address PR comments (code organization)

* Address PR comments (factory interface)

* fix test after rebase

* `remove`'s gone from the trait ...

... `DROP FUNCTION` will look for function name
in all available registries (udf, udaf, udwf).

`remove` may be necessary if UDaF and UDwF do not
get `simplify` method from #9304.

* Rename FunctionDefinition and export it ...

FunctionDefinition already exists, DefinitionStatement makes more sense.

* Update datafusion/expr/src/logical_plan/ddl.rs

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* Update datafusion/core/src/execution/context/mod.rs

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* Update datafusion/core/tests/user_defined/user_defined_scalar_functions.rs

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* Update datafusion/expr/src/logical_plan/ddl.rs

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* resolve part of follow up comments

* Qualified functions are not supported anymore

* update docs and todos

* fix clippy

* address additional comments

* Add sqllogicteset for CREATE/DROP function

* Add coverage for DROP FUNCTION IF EXISTS

* fix multiline error

* revert dialect back to generic in test ...

... as `create function` gets support in latest
sqlparser.

* fmt

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

metesynnada committed:
Testing and clippy refactors

metesynnada committed:
Merge remote-tracking branch 'upstream/main' into configuration

jayzhan211 committed:
Enable configurable display of partition sizes in the explain statement (#9474)

* quick fixes

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* remove one in load

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* fix others

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* revert

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* config

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* with_show_sizes

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

---------

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

metesynnada committed:
After merge corrections

metesynnada committed:
Merge remote-tracking branch 'upstream/main' into configuration

metesynnada committed:
Parquet feature fix

comphead committed:
Reduce casts for LEAD/LAG (#9468)

* Reduce casts for LEAD/LAG

* uncomment test

* fix test

* test

metesynnada committed:
On datafusion-cli register COPY statements

metesynnada committed:
Correcting a test

comphead committed:
[CI build] fix chrono suggestions (#9486)

* [CI] fix chrono suggestions

* fmt

* clippy

* fix examples

alamb committed:
Make regex dependency optional in datafusion-functions, add CI checks for function packages (#9473)

* Make regex dependency optional in datafusion-functions

* Add missing cargo-check tests in CI

* make it optional

* remove uncessary core expressions

* add missing dependency

* format toml

metesynnada committed:
Merge remote-tracking branch 'upstream/main' into configuration

viirya committed:
fix: coalesce function should return correct data type (#9459)

* fix: Remove supported coalesce types

* Use comparison_coercion

* Fix test

* Fix

* Add comment

* More

* fix

comphead committed:
LEAD/LAG calculate default value once (#9485)

* LEAD/LAG calculate default value once

* refmt

waynexia committed:
chore: simplify the return type of `validate_data_types()` (#9491)

Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

comphead committed:
minor: use arrow-rs casting (#9500)

dependabot[bot] committed:
chore(deps): update substrait requirement from 0.25.1 to 0.27.0 (#9502)

Updates the requirements on [substrait](https://github.com/substrait-io/substrait-rs) to permit the latest version.
- [Release notes](https://github.com/substrait-io/substrait-rs/releases)
- [Changelog](https://github.com/substrait-io/substrait-rs/blob/main/CHANGELOG.md)
- [Commits](https://github.com/substrait-io/substrait-rs/compare/v0.25.1...v0.27.0)

---
updated-dependencies:
- dependency-name: substrait
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jonahgao committed:
fix: `generate_series` and `range` panic on edge cases (#9503)

* fix: `generate_series` and `range` panic on edge cases

* fix comment

* avoid casting to i128

* add tests

* use target_pointer_width

* remove target_pointer_width

mustafasrepo committed:
Fix undeterministic behaviour of schema nullability of lag window query (#9508)

* Initial commit

* Update comment, remove leftovers

* Add test

* Update datafusion/sqllogictest/test_files/window.slt

Co-authored-by: comphead <comphead@users.noreply.github.com>

---------

Co-authored-by: comphead <comphead@users.noreply.github.com>

Tangruilin committed:
Add `to_unixtime` function (#9077)

* [task #5568] add_to_unixtime_function

Signed-off-by: tangruilin <tang.ruilin@foxmail.com>

* Fix clippy

---------

Signed-off-by: tangruilin <tang.ruilin@foxmail.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

alamb committed:
Minor: fixed transformed state return (#9484)

SteveLauC committed:
test: port strpos test in physical_expr/src/functions to sqllogictest (#9439)

jayzhan211 committed:
Port ArrayHas family to `functions-array` (#9496)

* array has rewrite

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* first draft

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* rm dims

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* replace optimizer

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* remove proto and import udf

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* Remove unecessary dependency

* Add doc

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* fmt

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

---------

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

Weijun-H committed:
port array_empty and array_length (#9510)

jonahgao committed:
fix: `substr_index` not handling negative occurrence correctly (#9475)

* fix: `substr_index` not handling negative occurrence correctly

* format test

* add test

* add more tests

Ted-Jiang committed:
[minor] extract collect file statistics method and add doc (#9490)

* [minor] extract collect file statistics method and add doc

* fix clippy

* fix doc

korowa committed:
test: sqllogictests for multiple tables join (#9480)

Lordworms committed:
Add support for ignore nulls for LEAD, LAG in WindowAggExec (#9498)

* adding prototype

a

* fix clippy

* modify test

* fix clippy

* fix:clippy

* optimize code

* change tests

caicancai committed:
Minior: Improve log expr description (#9516)

* Minior: Improve log expr description

* Minior: Improve log expr description

Weijun-H committed:
port flatten (#9523)

my-vegetable-has-exploded committed:
feat: Add projection to HashJoinExec. (#9236)

* Add projection to HashJoinExec.

* fix hashjoin display

* pushdown coalesce batch & fix try_new

* fix slt

* fix test

* fmt & clippy

* fix into_proto

* rm coalesce_batches_pushdown & add one pushdown before CoalesceBatches

* fix slt.

* fmt & clippy

* apply projection to equivalence_properties and output_ordering.

* fix proto.

* fix merge.

* Add more comments and some tests.

* fix test & projection stats

* use collect_columns & fix schema usage

milenkovicm committed:
Add example for `FunctionFactory` (#9482)

* `FunctionFactory` usage example

* update test to use the same function factory

* Add entry to examples/README.md

* Add SessionContext::with_function_factory

* Update doc and example

* clippy

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

Omega359 committed:
Move date_part, date_trunc, date_bin functions to datafusion-functions (#9435)

* Move date_part, date_trunc, date_bin functions to datafusion-functions

* I do not understand why the logical plan changed but updating the explain text to reflect the change. The physical plan is unchanged.

* Fix fmt

* Improvements to remove datafusion-functions dependency from sq and physical-expr

* Fix function arguments for date_bin, date_trunc and date_part.

* Fix projection change. Add new test date_bin monotonicity

---------

Co-authored-by: Mustafa Akur <mustafa.akur@synnada.ai>
Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

tshauck committed:
fix: support two argument TRIM (#9521)

* fix: support two argument TRIM

* refactor: fix auto space remove

* test: fix test

jayzhan211 committed:
Remove physical expr of ListIndex and ListRange, convert to `array_element` and `array_slice` functions (#9492)

* remove physical range and index

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* remove proto

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

---------

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

SteveLauC committed:
feat: function name hints for UDFs (#9407)

* feat: function name hints for UDFs

* refactor: rebase fn to xxx_names()

* style: fix clippy

* style: fix clippy

* Add test

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

alamb committed:
Minor: Improve documentation for registering `AnalyzerRule` (#9520)

* Minor: Improve documentation for registering `AnalyzerRule`

* Apply suggestions from code review

Co-authored-by: comphead <comphead@users.noreply.github.com>

* Update datafusion/core/src/physical_optimizer/optimizer.rs

---------

Co-authored-by: comphead <comphead@users.noreply.github.com>

jayzhan211 committed:
Extend argument types for udf `return_type_from_exprs` (#9522)

* Extend argument types for udf return type function

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* rm incorrect assumption

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* possible empty types

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

---------

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

guojidan committed:
move make_array array_append array_prepend array_concat function to datafusion-functions-array crate (#9504)

* move array function

* fix rebase

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* cleanup to trigger rerun

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* split functions to different files

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* fix

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* fix conflict

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

* clippy

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>

---------

Signed-off-by: jayzhan211 <jayzhan211@gmail.com>
Co-authored-by: jayzhan211 <jayzhan211@gmail.com>

erenavsarogullari committed:
Port `StringToArray` to `function-arrays` subcrate (#9543)

* Issue-9497 - Port StringToArray to function-arrays

* Issue-9497 - Fix formatting issues

* Issue-9497 - Format expressions.md documentation

alamb committed:
Minor: remove `..` pattern matching in sql planner (#9531)

yyy1000 committed:
fix doc (#9542)

yyy1000 committed:
Port `struct` to datafusion-functions (#9546)

* initial port

* fix clippy

* update cargo in cli

* remove dependency

* resolve conflict

* cargo update in CLI

lewiszlw committed:
UDAF and UDWF support aliases (#9489)

* UDAF and UDWF support aliases

* Add tests for udaf and udwf aliases

* Fix clippy lint

tshauck committed:
docs: fix extraneous char (#9560)

mustafasrepo committed:
Initial commit (#9559)

yyy1000 committed:
Port `arrow_typeof` to datafusion-function (#9524)

* Port arrowtypeof

* fmt

* fix test case

* revert test change

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

backkem committed:
feat: Introduce convert Expr to SQL string API and basic feature (#9517)

* feat: convert Expr to SQL string

* fix: add license headers

* fix: make Unparser and Dialect public

* Update datafusion/sql/src/unparser/dialect.rs

* fmt

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

ozankabak committed:
Review

erenavsarogullari committed:
Port `ArraySort` to `function-arrays` subcrate (#9551)

* Issue-9550 - Port ArraySort to function-arrays subcrate

* Issue-9550 - Add test coverage on roundtrip_logical_plan

* Issue-9550 - Address review comments

metesynnada committed:
Review visited

metesynnada committed:
Merge remote-tracking branch 'upstream/main' into configuration

ozankabak commented:
Merged upstream.
